taskcat/taskcat-v9
# Lambda build

## Overview

adds a package command to handle lambda builds via docker. Has various differences to the v8 `-b` implementation:

* uses docker to package - previously we just zipped whatever was in the source foders, so dependencies had to be checked into the repo, and binary/platform specific deps required being built on a supported platform.
* recursive - previously submodule packages were not built, so if your project was used as a submodule you would need to check in the package zips.

The package command will check each source folder for the presence of a `Dockerfile` if it's present it will do a `docker build`, and then a `docker run` mounting the `packages/FunctionName/` path to `/output` folder in the container. 

Here's an example Dockerfile that does a `pip install` on a `requirements.txt` file in the source folder, and packages everything up into a zip:

```
FROM lambci/lambda:build-python3.7

COPY . .

RUN pip install -t . -r ./requirements.txt && \
    rm -rf *.dist-info *.pth && \
    rm Dockerfile requirements.txt

CMD zip -FS -r /output/lambda.zip ./
```

## Testing/Steps taken to ensure quality

tested by hand and added tests in the test suite. 

### Notes

Optional. Caveats, Alternatives, Other relevant information.

## Testing Instructions

 How to test this PR Start after checking out this branch (bulleted)
 * run unit tests
 * run `taskcat-v9 --debug package -c ./tests/data/lambda_build_with_submodules/.taskcat.yml -p ~/taskcat-v9/tests/data/lambda_build_with_submodules/`
 * check that package folders and zip files with the expected contents have been created in the `~/taskcat-v9/tests/data/lambda_build_with_submodules/` project and it's submodules
